moquette-0.16/broker/src/main/java/io/moquette/broker/subscriptions/Topic.java

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright (c) 2012-2018 The original author or authors
+ * ------------------------------------------------------
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Apache License v2.0 which accompanies this distribution.
+ *
+ * The Eclipse Public License is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * The Apache License v2.0 is available at
+ * http://www.opensource.org/licenses/apache2.0.php
+ *
+ * You may elect to redistribute this code under either of these licenses.
+ */
+
+package io.moquette.broker.subscriptions;
+
+import java.io.Serializable;
+import java.text.ParseException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class Topic implements Serializable, Comparable<Topic> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(Topic.class);
+
+    private static final long serialVersionUID = 2438799283749822L;
+
+    private final String topic;
+
+    private transient List<Token> tokens;
+
+    private transient boolean valid;
+
+    /**
+     * Factory method
+     *
+     * @param s the topic string (es "/a/b").
+     * @return the created Topic instance.
+     * */
+    public static Topic asTopic(String s) {
+        return new Topic(s);
+    }
+
+    public Topic(String topic) {
+        this.topic = topic;
+    }
+
+    Topic(List<Token> tokens) {
+        this.tokens = tokens;
+        List<String> strTokens = tokens.stream().map(Token::toString).collect(Collectors.toList());
+        this.topic = String.join("/", strTokens);
+        this.valid = true;
+    }
+
+    public List<Token> getTokens() {
+        if (tokens == null) {
+            try {
+                tokens = parseTopic(topic);
+                valid = true;
+            } catch (ParseException e) {
+                valid = false;
+                LOG.error("Error parsing the topic: {}, message: {}", topic, e.getMessage());
+            }
+        }
+
+        return tokens;
+    }
+
+    private List<Token> parseTopic(String topic) throws ParseException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 97-122 into a separate method.